A66-otel-stats.md

```diff
@@ -0,0 +1,569 @@
+
+The following call-outs are needed on the `CallAttemptTracer` -
+
+*   When a new message is sent/received. The message should be in its compressed
+    form.
+*   When the trailing metadata/status is received for the attempt. Receipt of
+    this indicates that the attempt has ended. Implementations may choose to
+    delegate the responsibility of notifying the `CallTracer` about the attempt
+    end to the `CallAttemptTracer`.
+
+The following call-outs are needed on the `ServerCallTracer` -
+
+*   When initial metadata is received by the transport for a call. This
+    indicates the start time of a new call.
+*   When a new message is sent/received. The message should be in its compressed
+    form.
+*   When trailing metadata/status is sent. This call-out should be as close to
+    the transport as possible to be able to capture the total time of the call.
+
+Implementations should allow multiple call/attempt tracers to be registered to a
+single call since there could be multiple plugins registered. For example, there
+could be an OpenCensus and an OpenTelemetry stats plugin registered together. It
+should also allow multiple OpenTelemetry plugins to be registered providing the
+ability to configure the different plugins with different MeterProviders.
+
+A sample implementation of this approach is available in
+[gRPC Core](https://github.com/grpc/grpc/blob/v1.57.x/src/core/lib/channel/call_tracer.h).
+
+In grpc-java, a client interceptor is provided by the gRPC OTel plugin. This
+interceptor adds a `CallAttemptTracerFactory` to the client call. This factory
+is equivalent to the `CallTracer`. For each attempt, this factory is invoked to
+create a `ClientStreamTracer` analogous to `CallAttemptTracer` for each attempt.
+On the server-side, a `ServerStreamTracer.Factory` is used to create tracers
+analogous to `ServerCallTracer` for each incoming call.
+
+In grpc-go, similar to grpc-java, an interceptor is invoked per call. This
+interceptor is registered when the OTel Dial Option is passed in to the channel,
+and has access to a context scoped to the call. `StatsHandler` object owned by
+the channel gets call-outs for each event that happens on the lifetime of an
+attempt. Along with each call-out gets, a context object scoped to the attempt
+is passed in, making it equivalent to the functionality of the
+`CallAttemptTracer`. On the server side, a `StatsHandler` object gets call-outs
+similarly along with a server call scoped context object, to get
+`ServerCallTracer` equivalent functionality.
+
+### Language-Specific Details
+
+Each language implementation will provide an API for registering an
+OpenTelemetry plugin. Overall, the APIs should have the following capabilities -
+
+*   Allow installing multiple OpenTelemetry plugins.
+*   Allow setting a
+    [MeterProvider](https://opentelemetry.io/docs/specs/otel/metrics/api/#meterprovider)
+    on individual plugins. Implementations should require a MeterProvider being
```

Comment:
Setting a meter provider is optional based on default no-op behavior. Can you rephrase it to say "Implementations must provide an option to set MeterProvider" or something similar?

Reply:
Done